### DISCUSSION TITLE
Get the Nth subnet of an IPNetwork

### DISCUSSION THREAD

**Discussion by Aison0** [netaddr/netaddr#393] [Q&A]
Hello,

Is it possible to get the Nth subnet of an IPNetwork? Example:

`p = IPNetwork("10.0.0.0/8")`
`s = p.subnet(16)[124]`
`print(s) ` outputs 10.124.0.0/8

Thanks,

**Comment by jstasiak** [comment]
Hey, you're on the right track with using [`IPNetwork.subnet`](https://netaddr.readthedocs.io/en/latest/api.html#netaddr.IPNetwork.subnet). Since it returns an iterator you need to find a way to get the nth element yielded, for example using `more-itertools`' [`nth`](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.nth) or by converting the iterator to a list to then index it:

```
In [4]: network = IPNetwork("10.0.0.0/8")

In [5]: list(network.subnet(16))[124]
Out[5]: IPNetwork('10.124.0.0/16')
```

Caveat: converting to a list will use a lot of memory if you have a lot of subnets (likely to mostly be be an issue with IPv6).

**Comment by Aison0** [comment]
Yeah, converting to a list is very bad, especially when using on ipv6 subnets. But also using `nth` may be a problem not? Since it iterates until the target index is reached. So all intermediate values are also calculated, which is fairly not necessary. One could directly compute the requested subnet in one step.

The following code really takes a lot of time:
```
p = IPNetwork("2a02:168:200f::/48")
print(more_itertools.nth(p.subnet(127), 0xfffff))
```

Maybe the method IPNetwork.subnet should be extended with a start index parameter? There is already a count parameter to limit the number.

> :param count: (optional) number of consecutive IP subnets to be returned.
 :param index: (optional) index of first IP subnet to be returned.

**Reply by Aison0** [reply]
Just tested this one:
```
class IPNetwork2(IPNetwork):

	def subnet2(self, prefixlen, count=None, index=None, fmt=None):

          A generator that divides up this IPNetwork's subnet into smaller
          subnets based on a specified CIDR prefix.
  
          :param prefixlen: a CIDR prefix indicating size of subnets to be
              returned.
          :param count: (optional) number of consecutive IP subnets to be
              returned.
          :param index: (optional) first IP subnet to be returned.
          :return: an iterator containing IPNetwork subnet objects.

	if not 0 <= self.prefixlen <= self._module.width:
		raise ValueError('CIDR prefix /%d invalid for IPv%d!' \	 % (prefixlen, self._module.version))

	if not self.prefixlen <= prefixlen:
		#   Don't return anything.
		return

	#   Calculate number of subnets to be returned.
	width = self._module.width
	max_subnets = 2 ** (width - self.prefixlen) // 2 ** (width - prefixlen)

	if index is None:
		index = 0
	else:
		if not 0 <= index < max_subnets:
			raise ValueError('index outside of current IP subnet boundary!')

	if count is None:
		count = max_subnets - index
	else:		
		if not 1 <= count <= max_subnets - index:
			raise ValueError('count outside of current IP subnet boundary!')

	base_subnet = self._module.int_to_str(self.first)
	i = index
	while (i < count):
		subnet = self.__class__('%s/%d' % (base_subnet, prefixlen), self._module.version)
		subnet.value += (subnet.size * i)
		subnet.prefixlen = prefixlen
		i += 1
		yield subnet
```
This code runs many times faster:
`print(more_itertools.nth(p.subnet2(127, index=0xffff), 0))`

Maybe it would be even interesting to have a specialized function IPNetwork.get_subnet(self, prefixlen, index)

**Reply by jstasiak** [reply]
That's fair. May I ask what problem are you actually aiming to solve with this (getting the nth subnet of a given prefix in a network)?

**Reply by Aison0** [reply]
I'm auto-generating network/router configuration files with jinja2 and netaddr is used to compute the different subnets. The manages over 100 subnets with each of them having address pools etc... all these are computed with netaddr. 

Example for radvd.conf generation:
e.ipv6 is an instance of your IPNetwork (that I have extended here)

```
{% for e in entities %}
# {{e.name}} Network
interface {{e.intf}}
{
        IgnoreIfMissing on;
        AdvSendAdvert on;
        AdvManagedFlag on;
        AdvOtherConfigFlag on;

        prefix {{e.ipv6.get_subnet(58, -2)}}
        {
                AdvOnLink on;
                AdvAutonomous off;
        };
        prefix {{e.ipv6.get_subnet(58, -1)}}
        {
                AdvOnLink on;
                AdvAutonomous on;
        };

        RDNSS {{e.ipv6[1]}} {{e.ipv6[2]}}
        {
        };
};
{% endfor %} 

Maybe i'm going to add more useful methods ;-)
```
